File: app/desktop/src/main/services/updater.ts
import * as fs from "fs";
import * as path from "path";
import { app, net } from "electron";
import type { GitHubRelease, UpdateCheckResult } from "../../renderer/types";

const RELEASES_API =
  "https://api.github.com/repos/RolinShmily/SrP-CFG_ForCS2/releases";
const CACHE_DIR = path.join(app.getPath("userData"), "update-cache");
const CACHE_FILE = path.join(CACHE_DIR, "cache.json");
const CHECK_INTERVAL = 4 * 60 * 60 * 1000; // 4 hours
const HTTP_TIMEOUT = 8000;

interface UpdateCache {
  lastCheckTime?: number;
  dismissedVersion?: string;
  cachedReleases?: GitHubRelease[];
  /** Full releases from GitHub (all versions), used as fallback for history */
  cachedAllReleases?: GitHubRelease[];
}

function getCurrentVersion(): string {
  return app.getVersion() || "0.1.0";
}

function compareVersions(a: string, b: string): number {
  const pa = a.replace(/^v/, "").split(".").map(Number);
  const pb = b.replace(/^v/, "").split(".").map(Number);
  for (let i = 0; i < 3; i++) {
    if ((pa[i] || 0) > (pb[i] || 0)) return 1;
    if ((pa[i] || 0) < (pb[i] || 0)) return -1;
  }
  return 0;
}

function loadCache(): UpdateCache | null {
  try {
    if (!fs.existsSync(CACHE_FILE)) return null;
    return JSON.parse(fs.readFileSync(CACHE_FILE, "utf-8"));
  } catch {
    return null;
  }
}

function saveCache(cache: UpdateCache): void {
  try {
    fs.mkdirSync(CACHE_DIR, { recursive: true });
    fs.writeFileSync(CACHE_FILE, JSON.stringify(cache, null, 2));
  } catch {
    // cache write failure is non-critical
  }
}

function isDismissed(
  releases: GitHubRelease[],
  dismissedVersion?: string,
): boolean {
  if (!dismissedVersion || releases.length === 0) return false;
  return compareVersions(releases[0].tagName, dismissedVersion) <= 0;
}

interface GitHubAssetRaw {
  name: string;
}

interface GitHubReleaseRaw {
  tag_name: string;
  name: string;
  body: string;
  html_url: string;
  published_at: string;
  assets: GitHubAssetRaw[];
}

async function fetchAllReleases(): Promise<GitHubReleaseRaw[]> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), HTTP_TIMEOUT);
  try {
    const res = await net.fetch(`${RELEASES_API}?per_page=10`, {
      headers: { "User-Agent": "SrP-CFG-Installer" },
      signal: controller.signal,
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    return (await res.json()) as GitHubReleaseRaw[];
  } finally {
    clearTimeout(timer);
  }
}

function hasDesktopAssets(assets: GitHubAssetRaw[]): boolean {
  return assets.some(
    (a) =>
      /^SrP-CFG_Installer\.msi$/i.test(a.name) ||
      /^SrP-CFG_Portable\.zip$/i.test(a.name),
  );
}

function mapRelease(r: GitHubReleaseRaw): GitHubRelease {
  return {
    tagName: r.tag_name.replace(/^v/, ""),
    name: r.name || "",
    body: r.body || "",
    htmlUrl: r.html_url,
    publishedAt: r.published_at || "",
    hasDesktopAssets: hasDesktopAssets(r.assets || []),
  };
}

function filterNewer(
  releases: GitHubRelease[],
  currentVersion: string,
): GitHubRelease[] {
  return releases.filter(
    (r) => compareVersions(r.tagName, currentVersion) > 0,
  );
}

function buildResult(
  current: string,
  releases: GitHubRelease[],
): UpdateCheckResult {
  const hasDesktopUpdate = releases.some((r) => r.hasDesktopAssets);
  const hasPresetUpdate = releases.some((r) => !r.hasDesktopAssets);
  return {
    currentVersion: current,
    hasUpdate: releases.length > 0,
    hasDesktopUpdate,
    hasPresetUpdate,
    releases,
  };
}

export async function checkForUpdate(
  force = false,
): Promise<UpdateCheckResult> {
  const current = getCurrentVersion();
  const cache = loadCache();

  // Auto-check with cache throttle
  if (
    !force &&
    cache?.lastCheckTime &&
    Date.now() - cache.lastCheckTime < CHECK_INTERVAL
  ) {
    const allReleases = cache.cachedAllReleases || cache.cachedReleases || [];
    const newer = filterNewer(allReleases, current);
    if (isDismissed(newer, cache.dismissedVersion)) {
      return buildResult(current, []);
    }
    return buildResult(current, newer);
  }

  try {
    const raw = await fetchAllReleases();
    const allReleases = raw
      .map(mapRelease)
      .sort((a, b) => compareVersions(b.tagName, a.tagName));
    const newer = filterNewer(allReleases, current);

    saveCache({
      lastCheckTime: Date.now(),
      dismissedVersion: cache?.dismissedVersion,
      cachedReleases: newer,
      cachedAllReleases: allReleases,
    });

    // Auto-check respects dismissal
    if (!force && isDismissed(newer, cache?.dismissedVersion)) {
      return buildResult(current, []);
    }

    return buildResult(current, newer);
  } catch {
    // Network failure, use cache
    const allReleases = cache?.cachedAllReleases || cache?.cachedReleases || [];
    const newer = filterNewer(allReleases, current);
    if (!force && isDismissed(newer, cache?.dismissedVersion)) {
      return buildResult(current, []);
    }
    return buildResult(current, newer);
  }
}

export async function fetchUpdateHistory(): Promise<GitHubRelease[] | null> {
  try {
    const raw = await fetchAllReleases();
    const allReleases = raw
      .map(mapRelease)
      .filter((r) => compareVersions(r.tagName, "3.0.0") >= 0)
      .sort((a, b) => compareVersions(b.tagName, a.tagName));

    // Update cache with full releases for fallback
    const cache = loadCache() || {};
    cache.cachedAllReleases = allReleases;
    saveCache(cache);

    return allReleases;
  } catch {
    // Network failure, use cached full releases
    const cache = loadCache();
    const cached = (cache?.cachedAllReleases || [])
      .filter((r) => compareVersions(r.tagName, "3.0.0") >= 0)
      .sort((a, b) => compareVersions(b.tagName, a.tagName));
    return cached.length > 0 ? cached : null;
  }
}

export function dismissVersion(version: string): void {
  const cache = loadCache() || {};
  cache.dismissedVersion = version;
  saveCache(cache);
}

export async function getLatestVersion(): Promise<string> {
  const cache = loadCache();
  if (
    cache?.lastCheckTime &&
    Date.now() - cache.lastCheckTime < CHECK_INTERVAL &&
    cache.cachedAllReleases?.length
  ) {
    return cache.cachedAllReleases[0].tagName;
  }

  try {
    const raw = await fetchAllReleases();
    if (raw.length === 0) return getCurrentVersion();
    const latest = mapRelease(raw[0]);
    return latest.tagName;
  } catch {
    return cache?.cachedAllReleases?.[0]?.tagName ?? getCurrentVersion();
  }
}

File: app/desktop/src-tauri/src/services/updater.rs
//! GitHub Releases 更新检测壳层。
//!
//! 对应原 `app/desktop/src/main/services/updater.ts`：
//! - 网络请求（ureq）→ core `map_release` / `filter_newer` / `build_result` /
//!   `is_dismissed` / `is_cache_fresh` / `filter_at_least` / `sort_newest_first`
//! - 缓存读写（update-cache/cache.json）+ 4 小时节流

use std::fs;
use std::path::PathBuf;

use srp_cfg_core::{
    build_result, filter_at_least, filter_newer, is_cache_fresh, is_dismissed, map_release,
    sort_newest_first, Release, UpdateCheckResult,
};

use crate::ctx;

const RELEASES_API: &str = "https://api.github.com/repos/RolinShmily/SrP-CFG_ForCS2/releases";
const CHECK_INTERVAL_MS: u64 = 4 * 60 * 60 * 1000;
const HTTP_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(8);
const HISTORY_MIN_VERSION: &str = "3.0.0";

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct UpdateCache {
    pub last_check_time: Option<u64>,
    pub dismissed_version: Option<String>,
    pub cached_releases: Option<Vec<Release>>,
    pub cached_all_releases: Option<Vec<Release>>,
}

fn cache_dir() -> PathBuf {
    ctx::user_data_dir().join("update-cache")
}
fn cache_file() -> PathBuf {
    cache_dir().join("cache.json")
}

fn load_cache() -> Option<UpdateCache> {
    fs::read_to_string(cache_file())
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
}

fn save_cache(cache: &UpdateCache) {
    let _ = fs::create_dir_all(cache_dir());
    let _ = fs::write(cache_file(), serde_json::to_string_pretty(cache).unwrap_or_default());
}

fn now_ms() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

fn get_current_version() -> String {
    crate::ctx::try_handle()
        .map(|h| h.package_info().version.to_string())
        .unwrap_or_else(|| "0.0.0".to_string())
}

// ── 网络 ───────────────────────────────────────────────────────

#[derive(Debug, serde::Deserialize)]
struct GitHubAssetRaw {
    name: String,
}

#[derive(Debug, serde::Deserialize)]
struct GitHubReleaseRaw {
    tag_name: String,
    /// GitHub API 对无标题的 release 返回 null（如 v3.1.4 的 name 为 null）。
    #[serde(default)]
    name: Option<String>,
    /// GitHub API 对无正文的 release 返回 null（如 v3.1.4 的 body 为 null）；
    /// 若按 String 严格反序列化会导致整个列表解析失败（L2 遗留验收实测发现）。
    #[serde(default)]
    body: Option<String>,
    html_url: String,
    published_at: String,
    assets: Vec<GitHubAssetRaw>,
}

fn fetch_all_releases() -> Result<Vec<GitHubReleaseRaw>, String> {
    let agent = ureq::Agent::config_builder()
        .timeout_global(Some(HTTP_TIMEOUT))
        // ureq 3.3 默认 TLS provider 为 Rustls（即使未编译该 feature 也会在 https 时 panic）；
        // 本工程编译 native-tls（Windows SChannel），必须显式指定 provider。
        .tls_config(
            ureq::tls::TlsConfig::builder()
                .provider(ureq::tls::TlsProvider::NativeTls)
                .build(),
        )
        .build()
        .new_agent();
    let resp = agent
        .get(&format!("{RELEASES_API}?per_page=10"))
        .header("User-Agent", "SrP-CFG-Installer")
        .call()
        .map_err(|e| format!("HTTP {e}"))?;
    let body = resp
        .into_body()
        .read_to_string()
        .map_err(|e| e.to_string())?;
    serde_json::from_str(&body).map_err(|e| e.to_string())
}

fn map_raw(r: &GitHubReleaseRaw) -> Release {
    let asset_names: Vec<String> = r.assets.iter().map(|a| a.name.clone()).collect();
    map_release(
        &r.tag_name,
        r.name.as_deref().unwrap_or(""),
        r.body.as_deref().unwrap_or(""),
        &r.html_url,
        &r.published_at,
        &asset_names,
    )
}

// ── 核心检查（对应 TS `checkForUpdate`）────────────────────────

pub fn check_for_update(force: bool) -> UpdateCheckResult {
    let current = get_current_version();
    let cache = load_cache();

    // 自动检查 + 缓存节流
    if !force {
        if let Some(c) = &cache {
            if c.last_check_time.is_some_and(|t| {
                is_cache_fresh(Some(t), now_ms(), CHECK_INTERVAL_MS)
            }) {
                let all = c
                    .cached_all_releases
                    .clone()
                    .or_else(|| c.cached_releases.clone())
                    .unwrap_or_default();
                let newer = filter_newer(&all, &current);
                if is_dismissed(newer.first().map(|r| r.tag_name.as_str()), c.dismissed_version.as_deref()) {
                    return build_result(&current, &[]);
                }
                return build_result(&current, &newer);
            }
        }
    }

    match fetch_all_releases() {
        Ok(raw) => {
            let mut all: Vec<Release> = raw.iter().map(map_raw).collect();
            all = sort_newest_first(&all);
            let newer = filter_newer(&all, &current);
            save_cache(&UpdateCache {
                last_check_time: Some(now_ms()),
                dismissed_version: cache.as_ref().and_then(|c| c.dismissed_version.clone()),
                cached_releases: Some(newer.clone()),
                cached_all_releases: Some(all),
            });
            if !force {
                if is_dismissed(
                    newer.first().map(|r| r.tag_name.as_str()),
                    cache.as_ref().and_then(|c| c.dismissed_version.as_deref()),
                ) {
                    return build_result(&current, &[]);
                }
            }
            build_result(&current, &newer)
        }
        Err(_) => {
            // 网络失败，使用缓存
            let all = cache
                .as_ref()
                .and_then(|c| c.cached_all_releases.clone())
                .or_else(|| cache.as_ref().and_then(|c| c.cached_releases.clone()))
                .unwrap_or_default();
            let newer = filter_newer(&all, &current);
            if !force {
                if is_dismissed(
                    newer.first().map(|r| r.tag_name.as_str()),
                    cache.as_ref().and_then(|c| c.dismissed_version.as_deref()),
                ) {
                    return build_result(&current, &[]);
                }
            }
            build_result(&current, &newer)
        }
    }
}

// ── 更新历史 / 忽略 / 最新版本 ────────────────────────────────

pub fn fetch_update_history() -> Option<Vec<Release>> {
    match fetch_all_releases() {
        Ok(raw) => {
            let mut all: Vec<Release> = raw.iter().map(map_raw).collect();
            all = filter_at_least(&all, HISTORY_MIN_VERSION);
            all = sort_newest_first(&all);
            let mut cache = load_cache().unwrap_or(UpdateCache {
                last_check_time: None,
                dismissed_version: None,
                cached_releases: None,
                cached_all_releases: None,
            });
            cache.cached_all_releases = Some(all.clone());
            save_cache(&cache);
            Some(all)
        }
        Err(_) => {
            let cache = load_cache()?;
            let mut cached = filter_at_least(
                cache.cached_all_releases.as_deref().unwrap_or(&[]),
                HISTORY_MIN_VERSION,
            );
            cached = sort_newest_first(&cached);
            if cached.is_empty() {
                None
            } else {
                Some(cached)
            }
        }
    }
}

pub fn dismiss_version(version: &str) {
    let mut cache = load_cache().unwrap_or(UpdateCache {
        last_check_time: None,
        dismissed_version: None,
        cached_releases: None,
        cached_all_releases: None,
    });
    cache.dismissed_version = Some(version.to_string());
    save_cache(&cache);
}

pub fn get_latest_version() -> String {
    let cache = load_cache();
    if let Some(c) = &cache {
        if c.last_check_time.is_some_and(|t| is_cache_fresh(Some(t), now_ms(), CHECK_INTERVAL_MS))
            && c.cached_all_releases.as_ref().is_some_and(|v| !v.is_empty())
        {
            return c
                .cached_all_releases
                .as_ref()
                .and_then(|v| v.first())
                .map(|r| r.tag_name.clone())
                .unwrap_or_else(|| get_current_version());
        }
    }
    match fetch_all_releases() {
        Ok(raw) => {
            if raw.is_empty() {
                get_current_version()
            } else {
                map_raw(&raw[0]).tag_name
            }
        }
        Err(_) => cache
            .as_ref()
            .and_then(|c| c.cached_all_releases.as_ref())
            .and_then(|v| v.first())
            .map(|r| r.tag_name.clone())
            .unwrap_or_else(|| get_current_version()),
    }
}
